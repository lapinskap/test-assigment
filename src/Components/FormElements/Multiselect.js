import PropTypes from 'prop-types';
import React from 'react';
import {
  InputGroup,
  InputGroupAddon,
  FormGroup,
  Label,
  InputGroupText,
  Input,

} from 'reactstrap';
import ValidationMessage from '../Form/ValidationMessage';
import __ from '../../utils/Translations';

export default function Multiselect({
  id, label, validation, invalid, value, prefix, suffix, validateField, tooltip, errorMessage, onChange, options, disabled, afterLabel,
}) {
  const inputOptions = options.map(
    ({ value: optionValue, label: optionLabel }) => (
      <option
        key={optionValue}
        value={optionValue}
      >
        {optionLabel ? __(optionLabel) : optionValue}
      </option>
    ),
  );

  return (
    <FormGroup>
      <Label data-t1={`${id}Label`} for={id}>
        {label}
        &nbsp;
        {tooltip}
      </Label>
      {afterLabel ? <span className="ml-3">{afterLabel}</span> : null}
      <InputGroup>
        {prefix ? (
          <InputGroupAddon addonType="prepend">
            <InputGroupText>{prefix}</InputGroupText>
          </InputGroupAddon>
        ) : null}
        <Input
          data-t1={id}
          type="select"
          onChange={onChange}
          id={id}
          disabled={disabled}
          name={id}
          invalid={invalid}
          onBlur={(e) => {
            if (validateField) {
              validateField(id, e.target.value, validation);
            }
          }}
          multiple
          value={value || []}
        >
          {inputOptions}
        </Input>
        <ValidationMessage message={errorMessage} />
        {suffix ? (
          <InputGroupAddon addonType="append">
            <InputGroupText>{suffix}</InputGroupText>
          </InputGroupAddon>
        ) : null}
      </InputGroup>
    </FormGroup>
  );
}

Multiselect.propTypes = {
  id: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  onChange: PropTypes.func.isRequired,
  tooltip: PropTypes.node,
  value: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
  disabled: PropTypes.bool,
  afterLabel: PropTypes.node,
  errorMessage: PropTypes.string,
  validateField: PropTypes.func,
  prefix: PropTypes.string,
  suffix: PropTypes.string,
  invalid: PropTypes.bool,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    label: PropTypes.string,
  })).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  validation: PropTypes.array,
};

Multiselect.defaultProps = {
  id: '',
  label: '',
  tooltip: null,
  afterLabel: null,
  value: [],
  errorMessage: null,
  validateField: null,
  validation: null,
  prefix: '',
  suffix: '',
  invalid: false,
  disabled: false,
};
