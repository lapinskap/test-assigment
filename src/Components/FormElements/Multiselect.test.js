import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Multiselect from './Multiselect';
import '../../tests/setupTests';

describe('<Multiselect />', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with all props', () => {
    const wrapper = renderer
      .create(<Multiselect
        id="idmultiselect"
        label="Multiselect field"
        validation={['required']}
        value={[4]}
        prefix="Random prefix"
        suffix="Random prefix"
        validateField={() => {
        }}
        tooltip={<div>Tooltip</div>}
        invalid
        errorMessage="This field i required"
        onChange={() => {
        }}
        options={[{ value: 1 }, { value: 2 }]}
      />);

    expect(wrapper.toJSON())
      .toMatchSnapshot();
  });

  it('renders correctly without optional props', () => {
    const wrapper = renderer
      .create(<Multiselect
        onChange={() => {
        }}
        value={null}
        afterLabel={<div>test after label</div>}
        options={[{
          value: 1,
          label: 'Option 1',
        }, {
          value: 2,
          label: 'Option 2',
        }]}
      />);

    expect(wrapper.toJSON())
      .toMatchSnapshot();
  });

  it('it validate on blur', () => {
    const clickEvent = { target: { value: [1] } };
    const validateField = jest.fn((id, inputValue, validation) => {
      expect(id).toBe('test');
      expect(inputValue.length).toBe(1);
      expect(validation).toBe('required');
    });
    const wrapper = shallow(<Multiselect
      id="test"
      onChange={() => {
      }}
      value={null}
      validation="required"
      afterLabel={<div>test after label</div>}
      validateField={validateField}
      options={[{
        value: 1,
        label: 'Option 1',
      }, {
        value: 2,
        label: 'Option 2',
      }]}
    />);
    wrapper.find('Input').at(0).prop('onBlur')(clickEvent);
    expect(validateField).toHaveBeenCalledTimes(1);
  });
});
