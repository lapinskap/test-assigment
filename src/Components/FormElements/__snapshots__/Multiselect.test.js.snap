// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Multiselect /> renders correctly with all props 1`] = `
<div
  className="form-group"
>
  <label
    className=""
    data-t1="idmultiselectLabel"
    htmlFor="idmultiselect"
  >
    Multiselect field
     
    <div>
      Tooltip
    </div>
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        Random prefix
      </span>
    </div>
    <select
      aria-invalid={true}
      className="is-invalid form-control"
      data-t1="idmultiselect"
      disabled={false}
      id="idmultiselect"
      multiple={true}
      name="idmultiselect"
      onBlur={[Function]}
      onChange={[Function]}
      value={
        Array [
          4,
        ]
      }
    >
      <option
        value={1}
      >
        1
      </option>
      <option
        value={2}
      >
        2
      </option>
    </select>
    <div
      className="invalid-feedback"
      style={
        Object {
          "display": "block",
        }
      }
    >
      This field i required
    </div>
    <div
      className="input-group-append"
    >
      <span
        className="input-group-text"
      >
        Random prefix
      </span>
    </div>
  </div>
</div>
`;

exports[`<Multiselect /> renders correctly without optional props 1`] = `
<div
  className="form-group"
>
  <label
    className=""
    data-t1="Label"
    htmlFor=""
  >
    
     
  </label>
  <span
    className="ml-3"
  >
    <div>
      test after label
    </div>
  </span>
  <div
    className="input-group"
  >
    <select
      aria-invalid={false}
      className="form-control"
      data-t1=""
      disabled={false}
      id=""
      multiple={true}
      name=""
      onBlur={[Function]}
      onChange={[Function]}
      value={Array []}
    >
      <option
        value={1}
      >
        Option 1
      </option>
      <option
        value={2}
      >
        Option 2
      </option>
    </select>
  </div>
</div>
`;
